### ISSUE TITLE
Clang doesn't warn about out-of-range float to int cast

### ISSUE THREAD

**Issue by zmodem** [llvm/llvm-project#53107]
Consider: (https://godbolt.org/z/K3aWWK8a4)

```
unsigned f() {
    return (unsigned)4294967808.;
}
```

Since the value is out of range, the behaviour is undefined and LLVM will optimize the result to undef, which is probably not what the developer wanted.

Clang doesn't warn there. It would warn however, if the conversion were implicit:

```
unsigned g() {
    return 4294967808.;
}

<source>:6:12: warning: implicit conversion of out of range value from 'double' to 'unsigned int' is undefined [-Wliteral-conversion]
    return 4294967808.;
    ~~~~~~ ^~~~~~~~~~~
```

Perhaps it would make sense to warn in both cases?

**Comment by tbaederr** [comment]
GCC behaves the same. @AaronBallman do you have an opinion on this? Should maybe go into clang-tidy instead (if it isn't already)?

**Comment by zmodem** [comment]
Unless there's something I'm missing, I think a regular warning would be better. I think clang-tidy is more for expensive checks and/or checks where it's less clear that the code is wrong. In this case it seems the compiler can say with complete confidence that the code is bad.

**Comment by AaronBallman** [comment]
The trouble is: the cast is how you silence the implicit conversion warning, so if we begin to diagnose always, users who want to say "trust me compiler, I know what I'm doing" have no way to do so short of pragmas or globally silencing the diagnostic, which we typically try to avoid.

However, diagnosing unconditional undef is a pretty nice service to provide users. Because the cast doesn't change the undef behavior, I'm sort of inclined to say this situation should always be diagnosed (even with the cast) and that forcing the users to do something more onerous is better than allowing them to trivially silence the warning about their broken code.

I'd be willing to consider unconditionally warning on that if we had some measure of how chatty the diagnostic is in practice over some large corpus of code with the diagnostic enabled (several large projects, C and C++, of varying code hygiene). My intuition is that this diagnostic won't trigger with too much greater frequency and that each case will be a true positive, but it'd be good to test that intuition as best we can.
